### ISSUE TITLE
Array of components 

### ISSUE THREAD

**Issue by B-LIE** [SciML/ModelingToolkit.jl#2674]
**Is there an example of how to use arrays of components in the documentation❓**

<img width="444" alt="image" src="https://github.com/SciML/ModelingToolkit.jl/assets/54220415/eea1c76e-5c99-4a55-b5d2-298d01e6a5c3">

From a presentation, I've seen the the syntax:

```julia
@components begin
    resistor[1:n] = Resistor()
end
```
I don't know if this is a standard notation or not. I've tried to use it, but have problems when I add equations where the instances have fields, such...
```julia
@equations begin
    resistor[1].R ~ 3 
end
```
would not work. 

However, 
```julia
@components begin
    resistor_1 = Resistor()
end
#
@equations begin
    resistor_1.R ~ 3
end
```
would work.

**Comment by ChrisRackauckas** [comment]
@ven-k is the array form documented?

**Comment by ven-k** [comment]
Yes its the examples here: https://docs.sciml.ai/ModelingToolkit/stable/basics/MTKModel_Connector/#What-can-an-MTK-Model-definition-have?

The relevant bits (from ModelC example):
```julia
@mtkmodel ModelC begin
    @structural_parameters begin
        N = 2
    end
:
    @components begin
        model_a = ModelA(; k_array)
        model_array_a = [ModelA(; k = i) for i in 1:N]
        model_array_b = for i in 1:N
            k = i^2
            ModelA(; k)
        end
    end
:
end
```
For `@named model_c = ModelC()` ,the subcomponents can be then accessed as `model_c.model_array_a_1`, `model_c.model_array_a_2` and so on.

**Comment by B-LIE** [comment]
Thanks for answer! Useful, but it only addresses part of my question, and misses the most important part (marked with red):

Creation of vectors:

The example you refer to illustrates how a vector of models, `model_array_a` and `model_array_b` can be created. I missed that. Yingbo has used a third way, e.g., `model_array_c[1:N] = ModelA()` in his presentations. Perhaps this third way is only suitable when all model instantiations are equal...

Reference to vector elements:

You write:


The relevant bits (from ModelC example):

@mtkmodel ModelC begin
    @structural_parameters begin
        N = 2
    end
:
    @components begin
        model_a = ModelA(; k_array)
        model_array_a = [ModelA(; k = i) for i in 1:N]
        model_array_b = for i in 1:N
            k = i^2
            ModelA(; k)
        end
    end
:
end

For @named model_c = ModelC() ,the subcomponents can be then accessed as model_c.model_array_a_1, model_c.model_array_a_2 and so on.


This is useful. However, it doesn't answer my question, which was: how do I address the elements of, say, model_a within the ModelC definition. In other words, suppose I add:


...

@components begin

        model_a = ModelA(; k_array)
        model_array_a = [ModelA(; k = i) for i in 1:N]
        model_array_b = for i in 1:N
            k = i^2
            ModelA(; k)
        end
    end

 @equations begin
      for i in 1:N
          model_array_a[i].k ~ 3
       end
    end
:
end

OK – not a perfect example since `k` is a parameter; in my case it would be a variable.

What I have inserted in red above does not work. Should it be `model_array_a_i` instead? Or something else?

BR,
Bernt


________________________________

**Comment by B-LIE** [comment]
To make it simpler, consider my slightly modified code from the documentation, where my changes are highlighted in *red*:

using ModelingToolkit
using ModelingToolkit: t

@mtkmodel ModelA begin
    @parameters begin
        k
        k_array[1:2]
    end
end

@mtkmodel ModelB begin
    @parameters begin
        p1 = 1.0, [description = "Parameter of ModelB"]
        p2 = 1.0, [description = "Parameter of ModelB"]
    end
end

@mtkmodel ModelC begin
    # @icon "https://github.com/SciML/SciMLDocs/blob/main/docs/src/assets/logo.png"
    @constants begin
        c::Int = 1, [description = "Example constant."]
    end
    @structural_parameters begin
        f = sin
        N = 2
    end
    begin
        v_var = 1.0
    end
    @variables begin
        v(t) = v_var
        v_array(t)[1:2, 1:3]
        v_for_defaults(t)
    end
    @extend ModelB(; p1)
    @components begin
        model_a = ModelA(; k_array = [1, 2])
        model_array_a = [ModelA(; k = i) for i in 1:N]
        model_array_b = for i in 1:N
            k = i^2
            ModelA(; k)
        end
    end
    @equations begin
        model_a.k ~ f(v)
        model_array_a[1:N].k ~ f(v)
    end
    @defaults begin
        v_for_defaults => 2.0
    end
end
;

The addition of model_array_a[1:N].k ~ f(v) causes the error message below [same error message if I write .~ ]::

@mtkbuild modelc = ModelC()
type Array has no field k Stacktrace:  [1] getproperty  @ .\Base.jl:37<https://untitled+.vscode-resource.vscode-cdn.net/Base.jl:37> [inlined]  [2] #__ModelC__#74  @ C:\Users\Bernt\.julia\packages\UnPack\EkESO\src\UnPack.jl:1396 [inlined]  [3] __ModelC__(; name::Symbol, v_array::Nothing, p1::Nothing, N::Int64, model_array_a__k::Nothing, f::typeof(sin), v::Nothing, model_a__k_array::Nothing, v_for_defaults::Nothing, model_array_b__k::Nothing)  @ Main .\none:0<https://untitled+.vscode-resource.vscode-cdn.net/none:0>  [4] (::ModelingToolkit.Model{typeof(__ModelC__), Dict{Symbol, Any}})(; ***@***.***{name::Symbol})  @ ModelingToolkit C:\Users\Bernt\.julia\packages\ModelingToolkit\QP8aH\src\systems\model_parsing.jl:25  [5] top-level scope  @ C:\Users\Bernt\.julia\packages\ModelingToolkit\QP8aH\src\systems\abstractsystem.jl:1396

________________________________

**Comment by B-LIE** [comment]
(My color coding disappeared when transferring my e-mail to the GitHub system...)

**Comment by cstjean** [comment]
> For `@named model_c = ModelC()` ,the subcomponents can be then accessed as `model_c.model_array_a_1`, `model_c.model_array_a_2` and so on.

How can I iterate over that array of components? Eg. I can write

```julia
connect(model_c.model_array_a_1, other_component)
connect(model_c.model_array_a_2, other_component)
connect(model_c.model_array_a_3, other_component)
connect(model_c.model_array_a_4, other_component)
```
but that's terrible. Can I somehow write a for loop? 

[Context](https://discourse.julialang.org/t/component-based-model-for-n-reactors-connected-to-a-tank/119536/8?u=cstjean)

**Comment by cstjean** [comment]
This does not work for me. MWE:

```julia
using ModelingToolkit    # v9.48.0
using ModelingToolkit: t_nounits as t

@connector MyComponent begin
   @variables begin
      x(t)
   end
   @equations begin
      x ~ 10
   end
end

@mtkmodel Moo begin
   @components begin
      some_comp = MyComponent()
      comp_array = [MyComponent() for i in 1:4]
   end
end

@named moo = Moo()

moo.some_comp      # works
moo.comp_array_1   # works
moo.comp_array     # ArgumentError: System moo: variable comp_array does not exist
```
This is the same problem highlighted in @B-LIE's comment above (https://github.com/SciML/ModelingToolkit.jl/issues/2674#issuecomment-2082182657)

**Comment by ven-k** [comment]
The component-array is splat before adding them as subsystems. So `moo.comp_array` won't work. So they should be queried along with number `moo.comp_array_i`.

**Comment by cstjean** [comment]
Sounds good! Is there any reason why we couldn't have `comp_array = [comp_array_1, comp_array_2, ...]` automatically defined by the macro? Then the above would just work, instead of having to go through `getproperty` and `Symbol`.

**Comment by ven-k** [comment]
ODESystem constructor doesn't accept Vector{ODESystem} (for subsystem) as input.

Also, `[getproperty(moo, Symbol(:comp_array_, i)) for i in 1:4]` is meant for accessing the subsystems outside the macro.
The `connect` statement [above](https://github.com/SciML/ModelingToolkit.jl/issues/2674#issuecomment-2437405524), is placed inside macro and macro handles that.

```julia
 @mtkmodel Moo begin
          @components begin
             some_comp = MyComponent()
             comp_array = [MyComponent() for i in 1:4]
          end
           @equations begin
               [connect(comp_array[i],some_comp) for i in 1:4]
           end
   end

comp_array = [getproperty(moo, Symbol(:comp_array_, i)) for i in 1:4]
```
Maybe there can be a helper function to access it. But I fear it will have many non-obvious failing edge cases. If I can come up with a reliable one, I'll add.

**Comment by cstjean** [comment]
> ODESystem constructor doesn't accept Vector{ODESystem} (for subsystem) as input.

That looks like a complicated change indeed.

For context, I wanted to have a component with N connectors. In that case, the `connect` equation is outside of the `@mtkmodel`, and it does not work:

```julia
using ModelingToolkit
using ModelingToolkit: t_nounits as t

@mtkmodel MyConnector begin
    @variables begin
        x(t)
    end
end

@mtkmodel MilkTank begin
    @components begin
        connections = [MyConnector() for i in 1:4]
    end
end

@mtkmodel Farm begin
    @components begin
        m1 = MilkTank()
	m2 = MilkTank()
    end
    @equations begin
        # Not the most intuitive example...
        [connect(co1, co2) for (co1, co2) in zip(m1.connections, m2.connections)]...
    end
end

@named farm = Farm()  # System m1: variable connections does not exist
```
But on further investigation, stream connectors look like a better solution to my problem, so I don't need multiple connectors. Thank you for the input!
